src/3d/castlecomponentserialize.pas

```diff
@@ -170,6 +184,54 @@ implementation
   CastleFilesUtils, CastleUtils, CastleLog, CastleStringUtils, CastleClassUtils,
   CastleURIUtils, CastleVectors, CastleColors;
 
+{ TRegisteredComponent ----------------------------------------------------- }
+
+procedure TRegisteredComponent.ValidateComponentNameBase;
+begin
+  if SCharIs(FComponentNameBase, 1, ['0'..'9']) then
+    FComponentNameBase := 'Component' + FComponentNameBase;
+end;
+
+procedure TRegisteredComponent.ConstructComponentNameBase;
+begin
+  FComponentNameBase := ComponentClass.ClassName;
+
+  // remove common prefixes
+  if IsPrefix('TCastleUserInterface', FComponentNameBase, true) then
+    FComponentNameBase := PrefixRemove('TCastleUserInterface', FComponentNameBase, true)
+  else
+  if IsPrefix('TCastle', FComponentNameBase, true) then
+    FComponentNameBase := PrefixRemove('TCastle', FComponentNameBase, true)
+  else
+  if IsPrefix('T', FComponentNameBase, true) then
+    FComponentNameBase := PrefixRemove('T', FComponentNameBase, true);
+
+  // move 2D and 3D to back, as component name cannot start with a number
+  if IsPrefix('2D', FComponentNameBase, true) then
+    FComponentNameBase := PrefixRemove('2D', FComponentNameBase, true) + '2D';
+  if IsPrefix('3D', FComponentNameBase, true) then
+    FComponentNameBase := PrefixRemove('3D', FComponentNameBase, true) + '3D';
+
+  ValidateComponentNameBase;
+
+  // in case the replacements above made '', fix it (can happen in case of TCastleUserInterface)
+  if FComponentNameBase = '' then
+    FComponentNameBase := 'Group';
+end;
+
+function TRegisteredComponent.GetComponentNameBase: String;
+begin
+  if FComponentNameBase = '' then
+    ConstructComponentNameBase;
+  Result := FComponentNameBase;
+end;
+
+procedure TRegisteredComponent.SetComponentNameBase(const Value: String);
+begin
+  FComponentNameBase := Value;
+  ValidateComponentNameBase;
```

Comment:
It's a bit complicated that getter does some job (ConstructComponentNameBase), and setter does some job (ValidateComponentNameBase). I would much prefer to just have `ComponentNameBase` a simple field, and move the logic into a complete routine like `GetFinalComponentName`.